capture.go

```diff
@@ -149,18 +153,41 @@ func alert(exception error) error {
 		return true
 	})
 
+	// Run handlers in goroutines, so that if one handler is deadlocked
+	// it does not prevent others from running, or us from returning.
+	type captID struct {
+		CaptureProvider
+		CaptureID
+	}
+	captured := make(chan captID)
+	defer close(captured)
+
 	for provider, handler := range capture {
-		defer func() {
-			if r := recover(); r != nil {
-				log.Printf("failed to capture exception (%q): %+v", provider, r)
-			}
+		provider := provider
+		handler := handler
+		go func() {
+			defer func() {
+				if r := recover(); r != nil {
+					log.Printf("failed to capture exception (%q): %+v", provider, r)
+				}
+			}()
+
+			id := handler(exception, arg...)
+			captured <- captID{provider, id}
 		}()
+	}
 
-		id := handler(exception, arg...)
-		if id != "" {
-			e.id[provider] = id
+allProviders:
+	for len(e.id) < len(capture) {
+		select {
+		case cID := <-captured:
+			e.id[cID.CaptureProvider] = cID.CaptureID
+		case <-time.After(CaptureTimeout):
```

Comment:
You allow more time if there are more handlers.
I suggest having a timer object so that the CaptureTimeout applies to the total.

Reply:
Thanks for pointing it out.  Now using a single timer.  Also added a test.